snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java

```diff
@@ -0,0 +1,386 @@
+    }
+
+    // Filter to handle the edge case of '_' appearing as a wildcard that can't be remapped the way
+    // it can for predicates in SELECT statements.
+    schemas.removeIf(sc -> !schema.schemaName().equalsIgnoreCase(sc.schemaName()));
+    return !schemas.isEmpty();
+  }
+
+  @Override
+  public List<SnowflakeIdentifier> listDatabases() {
+    List<SnowflakeIdentifier> databases;
+    try {
+      databases =
+          connectionPool.run(
+              conn ->
+                  queryHarness.query(
+                      conn,
+                      "SHOW DATABASES IN ACCOUNT",
+                      DATABASE_RESULT_SET_HANDLER,
+                      (String[]) null));
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Failed to list databases");
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(e, "Interrupted while listing databases");
+    }
+    databases.forEach(
+        db ->
+            Preconditions.checkState(
+                db.type() == SnowflakeIdentifier.Type.DATABASE,
+                "Expected DATABASE, got identifier '%s'",
+                db));
+    return databases;
+  }
+
+  @Override
+  public List<SnowflakeIdentifier> listSchemas(SnowflakeIdentifier scope) {
+    StringBuilder baseQuery = new StringBuilder("SHOW SCHEMAS");
+    String[] queryParams = null;
+    switch (scope.type()) {
+      case ROOT:
+        // account-level listing
+        baseQuery.append(" IN ACCOUNT");
+        break;
+      case DATABASE:
+        // database-level listing
+        baseQuery.append(" IN DATABASE IDENTIFIER(?)");
+        queryParams = new String[] {scope.toIdentifierString()};
+        break;
+      default:
+        throw new IllegalArgumentException(
+            String.format("Unsupported scope type for listSchemas: %s", scope));
+    }
+
+    final String finalQuery = baseQuery.toString();
+    final String[] finalQueryParams = queryParams;
+    List<SnowflakeIdentifier> schemas;
+    try {
+      schemas =
+          connectionPool.run(
+              conn ->
+                  queryHarness.query(
+                      conn, finalQuery, SCHEMA_RESULT_SET_HANDLER, finalQueryParams));
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Failed to list schemas for scope '%s'", scope);
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(
+          e, "Interrupted while listing schemas for scope '%s'", scope);
+    }
+    schemas.forEach(
+        schema ->
+            Preconditions.checkState(
+                schema.type() == SnowflakeIdentifier.Type.SCHEMA,
+                "Expected SCHEMA, got identifier '%s' for scope '%s'",
+                schema,
+                scope));
+    return schemas;
+  }
+
+  @Override
+  public List<SnowflakeIdentifier> listIcebergTables(SnowflakeIdentifier scope) {
+    StringBuilder baseQuery = new StringBuilder("SHOW ICEBERG TABLES");
```

Comment:
@dennishuo I tried running this catalog implementation with Spark and it appears that this statement doesn't appear to show external iceberg tables for me.  I can run:

```
SHOW ICEBERG TABLES IN SCHEMA <DB>.<SCHEMA>;
```

And there are now results, but running:

```
SHOW TABLES IN SCHEMA <DB>.<SCHEMA>;
```

Correctly shows the Iceberg tables and I can see their storage and metadata locations.

Reply:
Right, this is a current design decision on the backend -- this catalog only lists "managed" Iceberg Tables right now, not External Iceberg Tables. Naturally, as currently Private Preview features we'll be happy to gather customer feedback accordingly.

Reply:
I think this is ok for now and we can revisit once external tables goes GA.